### PR TITLE
fix(postgresql): serialize slow-query metrics and add local recipe

### LINKED ISSUE #5350: fix(postgresql): serialize slow-query numeric fields
## Problem

`get_postgresql_slow_queries` returns PostgreSQL `numeric` columns as `Decimal` values. Tool results eventually pass through JSON serialization, where those values raise `TypeError: Object of type Decimal is not JSON serializable`.

This was reproduced while live-verifying all PostgreSQL tools for #5145.

## Expected behavior

Slow-query timing and cache-hit fields should be returned as JSON-serializable numbers while preserving the existing result shape.

## Reproduction

1. Start PostgreSQL with `pg_stat_statements` enabled.
2. Run a query above the slow-query threshold.
3. Call `get_postgresql_slow_queries` and serialize its result with `json.dumps`.

## Proposed fix

Convert the `numeric` timing and cache-hit fields to `float` when building the tool result and add a regression test that serializes a result containing `Decimal` rows.

Found while preparing #5338.

### PR DESCRIPTION
**Pull request by arnavpopley**
Part of #5145
Fixes #5350

#### Describe the changes you have made in this PR -

Verifies all six registered PostgreSQL tools against PostgreSQL 16 with `pg_stat_statements` enabled and adds a reproducible local Docker recipe. The recipe starts and seeds PostgreSQL, verifies connectivity, runs a real investigation against a slow query, isolates the demo from saved integrations, and cleans up its container, temporary store directory, and environment variables.

Live verification exposed #5350: PostgreSQL `numeric` timing fields arrive from psycopg2 as `Decimal` values and could not be serialized as JSON. The tool now returns those metrics as floats, with a regression test covering serialization.

### Demo/Screenshot for feature changes and bug fixes -

```
$ opensre integrations verify postgresql
SERVICE    │ SOURCE    │ STATUS   │ DETAIL
postgresql │ local env │ ✓ passed │ Connected to PostgreSQL 16.15
           │           │          │ target database: app_db.
```

The documented investigation exercised all six PostgreSQL tools against the seeded slow query.

---

## Code Understanding and AI Usage

**Did you use AI assistance (ChatGPT, Claude, Copilot, etc.) to write any part of this code?**
- [ ] No, I wrote all the code myself
- [x] Yes, I used AI assistance (continue below)

**If you used AI assistance:**
- [x] I have reviewed every single line of the AI-generated code
- [x] I can explain the purpose and logic of each function/component I added
- [x] I have tested edge cases and understand how the code handles them
- [x] I have modified the AI output to follow this project's coding standards and conventions

**Explain your implementation approach:**

The local recipe uses environment-backed credentials and a missing store path inside a temporary directory, so existing saved integrations cannot override the demo and no internal store JSON is written. The runtime fix converts only PostgreSQL `numeric` result fields to floats at the tool boundary, preserving the existing response shape while making it JSON-serializable.

Validation:
- `make lint`
- `make format-check`
- `make typecheck`
- `uv run python -m pytest tests/integrations/test_postgresql.py` (21 passed)

---

## Checklist before requesting a review
- [x] I have added proper PR title and linked to the issues
- [x] I have performed a self-review of my code
- [x] **I can explain the purpose of every function, class, and logic block I added**
- [x] I understand why my changes work and have tested them thoroughly
- [x] I have considered potential edge cases and how my code handles them
- [x] If it is a core feature, I have added thorough tests
- [x] My code follows the project's style guidelines and conventions

---

Note: Please check **Allow edits from maintainers** if you would like us to assist in the PR.

### CODE REVIEW

**Comment by muddlebee** [comment]
Please remove the embedded Python that directly constructs and writes integrations-store JSON. User-facing docs should not teach or depend on the internal store schema. Keep `OPENSRE_INTEGRATIONS_STORE_PATH` pointed at an empty temporary file so existing saved integrations cannot override the exported PostgreSQL demo variables. During teardown, remove that temporary file and unset its path and the demo PostgreSQL variables. Keep the PostgreSQL startup, seed, verify, investigation, and output sections unchanged.

File: docs/postgresql.mdx [hunk 1]
```diff
@@ -98,6 +98,120 @@ Restart PostgreSQL, then:
...
+
+```bash
+rm -f "$OPENSRE_INTEGRATIONS_STORE_PATH"
+```
+
+Teardown:
+
+```bash
+docker rm -f postgres-dev
+```
```
[the remark this answers is not in the document]
Comment (arnavpopley) [reply docs/postgresql.mdx:197]: Resolved -- teardown now unsets POSTGRESQL_HOST/PORT/DATABASE/USERNAME/PASSWORD/SSL_MODE. Verified live.

File: docs/postgresql.mdx [hunk 2]
```diff
@@ -98,6 +98,120 @@ Restart PostgreSQL, then:
...
+SERVICE    │ SOURCE    │ STATUS   │ DETAIL
+postgresql │ local env │ ✓ passed │ Connected to PostgreSQL 16.15
+           │           │          │ target database: app_db.
+```
+
+Register the local instance in the store. `opensre integrations setup postgresql` is interactive; for a scriptable demo, write directly into an isolated store file instead of `~/.opensre/integrations.json` -- `OPENSRE_INTEGRATIONS_STORE_PATH` is a genuine, first-class override (`config/constants/tenancy.py`), so this never touches your real config:
+
+```bash
+export OPENSRE_INTEGRATIONS_STORE_PATH=$(mktemp /tmp/opensre-postgresql-demo.XXXXXX)
+python3 <<'INNERPY'
```
[the remark this answers is not in the document]
Comment (arnavpopley) [reply docs/postgresql.mdx]: Flagging a disagreement rather than silently complying or ignoring: switching just this snippet to `uv run python` would make it inconsistent with the rest of this same doc, which uses bare `opensre integrations verify postgresql` / `opensre investigate` throughout (that convention predates my changes) -- and with all 6 already-merged sibling recipes (Jenkins/Dagster/Temporal/OpenSearch/MySQL/RocketChat), which all use bare `python3` for the identical store-registration snippet. AGENTS.mds `uv run python` guidance reads to me as the dev-workflow convention for running this repos own code; this snippet is pure stdlib (json/os/pathlib) with zero project dependencies, editing a JSON file that has nothing to do with the opensre package. If the intent is that every recipe across this whole batch should switch to `uv run` for both the CLI commands and the python snippets, I'm glad to do that consistently across all 7 docs in one pass -- just didn't want to make this one file inconsistent with its own neighboring commands and the merged precedent without confirming that's actually wanted first.

**Comment by arnavpopley** [comment]
@greptile review

File: docs/postgresql.mdx [hunk 3]
```diff
@@ -98,6 +98,122 @@ Restart PostgreSQL, then:
...
+performance problem -- no other slow queries, no locks, no long
+transactions, no bloat, healthy cache/connection counts.
+```
+
+All 6 registered tools were exercised in this single turn.
+
+Remove the demo store file when done (no edit to your real store needed):
+
+```bash
+rm -f "$OPENSRE_INTEGRATIONS_STORE_PATH"
```
[the remark this answers is not in the document]
Comment (arnavpopley) [reply docs/postgresql.mdx]: Resolved -- added the unset. Verified live that a subsequent opensre command correctly falls back to the default store path after rm + unset.

**Comment by arnavpopley** [comment]
@greptile review

File: docs/postgresql.mdx [hunk 4]
```diff
@@ -98,6 +98,123 @@ Restart PostgreSQL, then:
...
+unset OPENSRE_INTEGRATIONS_STORE_PATH
+```
+
+Teardown:
+
+```bash
+docker rm -f postgres-dev
+unset POSTGRESQL_HOST POSTGRESQL_PORT POSTGRESQL_DATABASE
+unset POSTGRESQL_USERNAME POSTGRESQL_PASSWORD POSTGRESQL_SSL_MODE
```
[the remark this answers is not in the document]
Comment (arnavpopley) [reply docs/postgresql.mdx:196]: Disagree, for consistency: this repo already has this exact `unset` pattern in the merged/pushed sibling recipe (docs/mysql.mdx teardown -- `unset MYSQL_HOST MYSQL_PORT MYSQL_DATABASE MYSQL_USERNAME MYSQL_PASSWORD MYSQL_SSL_MODE`), which passed Greptile 5/5 without this concern being raised. All of these recipes are scoped to a fresh scratch shell for a one-off local demo (temp store file, `--rm` container), not a session where the reader already has real POSTGRESQL_* creds exported for a live production integration. Snapshotting and restoring prior values would add real complexity (capture-if-set, restore-or-unset logic) to every one of these docs for an edge case none of them currently guard against. Leaving this open for a human call on whether that consistency-wide change is wanted -- happy to apply it repo-wide if so.

**Comment by arnavpopley** [comment]
Addressed in 291b1b2d3: removed the embedded `python3 <<INNERPY` store-construction block and replaced it with `OPENSRE_INTEGRATIONS_STORE_PATH` pointed at an explicit empty, valid store (`{"version": 2, "integrations": []}`), so the doc no longer teaches the internal store schema and the `POSTGRESQL_*` env vars are the sole source of connection info. Teardown now removes the temp file and unsets `OPENSRE_INTEGRATIONS_STORE_PATH` plus all `POSTGRESQL_*` demo vars. Also, while re-verifying this recipe live, found and fixed a real bug this surfaced: `get_postgresql_slow_queries` cast timing columns to `::numeric` in SQL, which psycopg2 returns as `Decimal` -- `opensre investigate` crashed with `TypeError: Object of type Decimal is not JSON serializable` on the final CLI output. Fixed by casting to `float`, with a regression test that reproduces the crash pre-fix and passes post-fix. Verified live end-to-end -- all 6 registered tools exercised, no crash, real store untouched.

**Comment by arnavpopley** [comment]
@greptile review

**Comment by arnavpopley** [comment]
Re-review from OpenSRE Development Bot in Slack:

### :white_check_mark: PR #5338 — docs(postgresql): add local verification recipe + fix Decimal serialization crash
**+79/-33 net · docs/postgresql.mdx, integrations/postgresql/__init__.py, tests/integrations/test_postgresql.py**

**Bug fix — the important part:**
get_slow_queries casts timing columns to ::numeric in SQL, which psycopg2 returns as Python Decimal. opensre investigate then crashes with TypeError: Object of type Decimal is not JSON serializable on final output. Fixed by casting all 6 timing columns (total_time_ms, mean_time_ms, min_time_ms, max_time_ms, stddev_time_ms, cache_hit_percent) to float in the query.

This is a live crash on any PostgreSQL instance with pg_stat_statements enabled — which is basically every production Postgres. High real-world impact.

**Regression test:**
- Mocks the cursor to return Decimal rows (matching actual psycopg2 behavior)
- Asserts float types in output
- Asserts json.dumps(result) doesn't raise
- Confirmed red on pre-fix code, green post-fix :white_check_mark:

This is exactly the right test to write here — it pins the behavior against real psycopg2 types, not just the fixed code path.

**Docs recipe:**
- postgres:16 with pg_stat_statements enabled
- Bounded readiness loop, seeded orders table + pg_sleep(2) slow query to exercise the tool
- Registration via empty valid store file + env vars — correctly avoids teaching the internal store JSON schema (good catch from muddlebee's review)
- All 6 registered tools called in the investigate run
- Correctly triages the seeded pg_sleep as a non-incident (proves the tool reasons correctly, not just that it runs)
- Complete teardown: container, temp file, OPENSRE_INTEGRATIONS_STORE_PATH, and all 6 POSTGRESQL_* vars unset

The -33 lines are the old embedded python3 snippet being replaced — a clear improvement in both safety and readability.

**Comment by muddlebee** [comment]
@greptile review
